crates/ide/src/annotations.rs

```diff
@@ -0,0 +1,139 @@
+) -> Vec<Annotation> {
+    let mut annotations = Vec::default();
+
+    if annotate_runnables {
+        for runnable in runnables(db, file_id) {
+            if !matches!(runnable.kind, RunnableKind::Bin) || !binary_target {
+                continue;
+            }
+
+            let action = runnable.action();
+            let range = runnable.nav.full_range;
+
+            if run {
+                annotations.push(Annotation {
+                    range,
+                    // FIXME: This one allocates without reason if run is enabled, but debug is disabled
+                    kind: AnnotationKind::Runnable { debug: false, runnable: runnable.clone() },
+                });
+            }
+
+            if action.debugee && debug {
+                annotations.push(Annotation {
+                    range,
+                    kind: AnnotationKind::Runnable { debug: true, runnable },
+                });
+            }
+        }
+    }
+
+    file_structure(&db.parse(file_id).tree())
+        .into_iter()
+        .filter(|node| {
+            matches!(
+                node.kind,
+                SymbolKind::Trait
+                    | SymbolKind::Struct
+                    | SymbolKind::Enum
+                    | SymbolKind::Union
+                    | SymbolKind::Const
+            )
+        })
+        .for_each(|node| {
+            if annotate_impls && node.kind != SymbolKind::Const {
+                annotations.push(Annotation {
+                    range: node.node_range,
+                    kind: AnnotationKind::HasImpls {
+                        position: FilePosition { file_id, offset: node.navigation_range.start() },
+                        data: None,
+                    },
+                });
+            }
+
+            if annotate_references {
+                annotations.push(Annotation {
+                    range: node.node_range,
+                    kind: AnnotationKind::HasReferences {
+                        position: FilePosition { file_id, offset: node.navigation_range.start() },
+                        data: None,
+                    },
+                });
+            }
+        });
+
+    if annotate_method_references {
+        annotations.extend(find_all_methods(db, file_id).into_iter().map(|method| Annotation {
+            range: method.range,
+            kind: AnnotationKind::HasReferences {
+                position: FilePosition { file_id, offset: method.range.start() },
+                data: None,
+            },
+        }));
+    }
+
+    annotations
+}
+
+pub(crate) fn resolve_annotation(db: &RootDatabase, annotation: Annotation) -> Annotation {
+    match annotation.kind {
+        AnnotationKind::HasImpls { position, .. } => {
+            let data = goto_implementation(db, position).map(|range| range.info);
+
+            Annotation {
+                range: annotation.range,
+                kind: AnnotationKind::HasImpls { position, data },
+            }
+        }
```

Comment:
I think you can do something more cute here:

```suggestion
pub(crate) fn resolve_annotation(db: &RootDatabase, annotation: mut Annotation) -> Annotation {
    match &mut annotation.kind {
        AnnotationKind::HasImpls { position, data } => {
            *data = goto_implementation(db, position).map(|range| range.info);
        }
```